### PR TITLE
Decompile gui initialisation

### PR DESCRIPTION
**Pull request by marijnvdwerf**
Should be a match, but my code likely isn't what's expected.

### CODE REVIEW

File: src/openloco/gui.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                window_type::toolbar_time,
+                uiWidth - 140,
+                uiHeight - 27,
+                140,
+                28,
+                0x32,
+                (void*)0x4fa098);
+            window->widgets = (ui::widget*)0x509d5c;
+            window->enabled_widgets = (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7);
+            window->pad_all[0x854] = 0;
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: Not really padding if it gets set.

File: src/openloco/objects/interface_skin_object.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+namespace openloco
+{
+    struct interface_skin_object
+    {
+        uint8_t pad_00[6];
+        uint8_t colours_06[2];
+        uint8_t pad_08[0x12 - 8];
+        uint8_t colours_12[6];
```
Comment (duncanspumpkin) [inline src/openloco/objects/interface_skin_object.h]: Are you able to get any better name? Looks like one is used during tutorials.

Comment (duncanspumpkin) [reply src/openloco/objects/interface_skin_object.h]: Actually maybe not. Scratch that

File: src/openloco/gui.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+        // viewport_create
+        registers regs;
+        regs.edx = 0x17FF17FF;
+        regs.eax = (window->y << 16) | window->x;
+        regs.ebx = (window->height << 16) | window->width;
+        regs.ecx = 0x1e00000;
+        regs.esi = (uint32_t)window;
+        addr<0x00e3f0b8, int32_t>() = 0; // gCurrentRotation
+        call(0x004ca2d0, regs);
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: Should just create a wrapper for this then you can have a function with zero use of registers.

File: src/openloco/gui.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+        _mainWindowWidgets[0].bottom = uiHeight;
+        _mainWindowWidgets[0].right = uiWidth;
+        auto window = openloco::ui::windowmgr::create_window(
+            window_type::main,
+            0,
+            0,
+            uiWidth,
+            uiHeight,
+            0x1,
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: Split this into component flags (1<<0). Same for all the other ones.

File: src/openloco/viewportmgr.cpp [hunk 5]
```diff
@@ -16,4 +16,16 @@ namespace openloco::ui::viewportmgr
...
         return *arr;
     }
+
+    void create(window* window, int16_t x, int16_t y, uint16_t width, uint16_t height)
+    {
+        registers regs;
+        regs.edx = 0x17FF17FF;
+        regs.eax = (y << 16) | x;
+        regs.ebx = (height << 16) | width;
+        regs.ecx = 0x1e00000;
```
Comment (duncanspumpkin) [inline src/openloco/viewportmgr.cpp]: `cl` is `zoom`. No idea what the `0x1E0` is about.

File: src/openloco/viewportmgr.cpp [hunk 6]
```diff
@@ -16,4 +16,16 @@ namespace openloco::ui::viewportmgr
...
+        regs.esi = (uint32_t)window;
+        addr<0x00e3f0b8, int32_t>() = 0; // gCurrentRotation?
```
Comment (duncanspumpkin) [inline src/openloco/viewportmgr.cpp]: move this back to the main window function.

File: src/openloco/window.h [hunk 7]
```diff
@@ -18,23 +31,37 @@ namespace openloco::ui
...
+                uint8_t pad_840[0x846 - 0x840];
+                uint16_t var_846;
+                uint8_t pad_848[0x854 - 0x848];
+                uint16_t var_854;
+                uint16_t var_856;
+                uint8_t pad_858[0x85A - 0x858];
                 uint16_t var_85A;
                 uint8_t pad_85C[0x870 - 0x85C];
                 uint16_t var_870;
                 uint8_t pad_872[0x882 - 0x872];
```
Comment (duncanspumpkin) [inline src/openloco/window.h]: 876 word is the `thing` that the viewport is looking at. 0xFFFF for not looking at a `thing`.

File: src/openloco/viewportmgr.cpp [hunk 8]
```diff
@@ -16,4 +16,16 @@ namespace openloco::ui::viewportmgr
         auto arr = (std::array<viewport*, max_viewports>*)_viewports.get();
         return *arr;
     }
+
+    void create(window* window, int16_t x, int16_t y, uint16_t width, uint16_t height)
+    {
+        registers regs;
+        regs.edx = 0x17FF17FF;
```
Comment (duncanspumpkin) [inline src/openloco/viewportmgr.cpp]: edx is either the `thing_id` or the x, y tile coordinates of the viewport. There is also flags encoded into bit 31 and bit 30. Bit 31 is set when a `thing_id` is being passed. bit 30 does something im not entirely sure to do with the zoom.

File: src/openloco/viewportmgr.cpp (hunk 5 above)
Comment (duncanspumpkin) [reply src/openloco/viewportmgr.cpp]: upper word of `ecx` is the z play space coordinate of the viewport.

File: src/openloco/gui.cpp [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        else
+        {
+
+            window = openloco::ui::windowmgr::create_window(
+                window_type::toolbar_top,
+                0,
+                0,
+                uiWidth,
+                28,
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: 27

File: src/openloco/gui.cpp [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                window->colours[2] = skin->colours_12[2];
+                window->colours[3] = skin->colours_12[3];
+            }
+
+            window = openloco::ui::windowmgr::create_window(
+                window_type::toolbar_player_info,
+                0,
+                uiHeight - 27,
+                140,
+                28,
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: 27

File: src/openloco/gui.cpp [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                window->colours[0] = colour::translucent(skin->colours_12[4]);
+                window->colours[1] = colour::translucent(skin->colours_12[4]);
+            }
+
+            window = openloco::ui::windowmgr::create_window(
+                window_type::toolbar_time,
+                uiWidth - 140,
+                uiHeight - 27,
+                140,
+                28,
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: 27

File: src/openloco/gui.cpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+            if (openloco::tutorial::state() != tutorial::tutorial_state::none)
+            {
+
+                window = openloco::ui::windowmgr::create_window(
+                    window_type::tutorial,
+                    140,
+                    uiHeight - 27,
+                    uiWidth - 280,
+                    28,
```
Comment (duncanspumpkin) [inline src/openloco/gui.cpp]: 27

**Comment by IntelOrca** [comment]
> I'm not sure if all of the access to memory should be made into separate loco_global global variables. I leave that to @IntelOrca to weigh in on.

Defining structs makes it a bit cleaner, but that can be done at a later date.

**Comment by marijnvdwerf** [comment]
@duncanspumpkin Can you do the viewport function fixes?

**Comment by duncanspumpkin** [comment]
Sure will do it in a few hours time

File: src/openloco/viewportmgr.cpp [hunk 13]
```diff
@@ -1,5 +1,6 @@
 #include "viewportmgr.h"
 #include "interop/interop.hpp"
+#include "things\thing.h"
```
Comment (AaronVanGeffen) [inline src/openloco/viewportmgr.cpp]: This won't work on Linux or macOS. ;)

**Pull request merged**